src/algebra/add_torsor.lean

```diff
@@ -0,0 +1,191 @@
+for multiplicative torsors (currently mathlib only develops the theory
+of group actions for multiplicative group actions).  The variable `G`
+is an explicit rather than implicit argument to lemmas because
+otherwise the elaborator sometimes has problems inferring appropriate
+types and type class instances.
+
+## References
+
+* https://en.wikipedia.org/wiki/Principal_homogeneous_space
+* https://en.wikipedia.org/wiki/Affine_space
+
+-/
+
+/-- Type class for the `+ᵥ` notation. -/
+class has_vadd (G : Type*) (P : Type*) :=
+(vadd : G → P → P)
+
+/-- Type class for the `-ᵥ` notation. -/
+class has_vsub (G : Type*) (P : Type*) :=
+(vsub : P → P → G)
+
+infix ` +ᵥ `:65 := has_vadd.vadd
+infix ` -ᵥ `:65 := has_vsub.vsub
+
+section prio
+set_option default_priority 100 -- see Note [default priority]
+set_option old_structure_cmd true
+/-- Type class for additive monoid actions. -/
+class add_action (G : Type*) (P : Type*) [add_monoid G] extends has_vadd G P :=
+(zero_vadd' : ∀ p : P, (0 : G) +ᵥ p = p)
+(vadd_assoc' : ∀ (g1 g2 : G) (p : P), g1 +ᵥ (g2 +ᵥ p) = (g1 + g2) +ᵥ p)
+
+/-- An `add_torsor G P` gives a structure to the nonempty type `P`,
+acted on by an `add_group G` with a transitive and free action given
+by the `+ᵥ` operation and a corresponding subtraction given by the
+`-ᵥ` operation. In the case of a vector space, it is an affine
+space. -/
+class add_torsor (G : Type*) (P : Type*) [add_group G] extends add_action G P, has_vsub G P :=
+(nonempty : nonempty P)
+(vsub_vadd' : ∀ (p1 p2 : P), (p1 -ᵥ p2 : G) +ᵥ p2 = p1)
+(vadd_vsub' : ∀ (g : G) (p : P), g +ᵥ p -ᵥ p = g)
+end prio
+
+/-- An `add_group G` is a torsor for itself. -/
+instance add_group_is_add_torsor (G : Type*) [add_group G] :
+  add_torsor G G :=
+{ vadd := has_add.add,
+  vsub := has_sub.sub,
+  zero_vadd' := zero_add,
+  vadd_assoc' := λ a b c, (add_assoc a b c).symm,
+  nonempty := ⟨0⟩,
+  vsub_vadd' := sub_add_cancel,
+  vadd_vsub' := add_sub_cancel }
+
+namespace add_torsor
+
+section general
+
+variables (G : Type*) {P : Type*} [add_group G] [S : add_torsor G P]
+include S
+
+/-- Adding the zero group element to a point gives the same point. -/
+@[simp] lemma zero_vadd (p : P) : (0 : G) +ᵥ p = p :=
+zero_vadd' p
+
+/-- Adding two group elements to a point produces the same result as
+adding their sum. -/
+lemma vadd_assoc (g1 g2 : G) (p : P) : g1 +ᵥ (g2 +ᵥ p) = (g1 + g2) +ᵥ p :=
+vadd_assoc' g1 g2 p
+
+/-- Adding the result of subtracting from another point produces that
+point. -/
+@[simp] lemma vsub_vadd (p1 p2 : P) : (p1 -ᵥ p2 : G) +ᵥ p2 = p1 :=
+vsub_vadd' p1 p2
+
+/-- Adding a group element then subtracting the original point
+produces that group element. -/
+@[simp] lemma vadd_vsub (g : G) (p : P) : g +ᵥ p -ᵥ p = g :=
+vadd_vsub' g p
+
+/-- If the same point added to two group elements produces equal
+results, those group elements are equal. -/
+lemma vadd_cancel_right (g1 g2 : G) (p : P) (h : g1 +ᵥ p = g2 +ᵥ p) : g1 = g2 :=
```

Comment:
In `add_right_cancel` the arguments corresponding to `g1`, `g2`, and `p` are implicit.

Reply:
Made the arguments corresponding to `g1` and `g2` implicit. In several places in this file I use this lemma via `apply`, which requires naming `p` explicitly as an argument; rather than using `@` in all those places, I think that suggests it's right for `p` to be an explicit argument here.